### PR TITLE
Initialize Kimi-K2 top-k output buffers

### LINKED ISSUE #339: Kimi-K2 top-k can return out-of-range expert indices when default output buffers are uninitialized
## Summary

`warp_kimi_sigmoid_top_8_static_v2` allocates its default output tensors with
`torch::empty()` before passing them to the bundled
`kimi_k2_sigmoid_top_k_routed_scaled_f32` cubin.

On NVIDIA B200, identical inputs produced different selected expert sets across
repeated calls. Some executions also returned expert indices outside the valid
range. Replacing the default output allocations with `torch::zeros()` eliminated
both behaviors in our tests.

## Affected code

Tested at commit `96991a970c6287f9a896810941e59578ac257cb3`. The same allocation
pattern is present on the current `main` branch in
`tutel/custom/custom_kernel.cpp`:

```cpp
auto top_v_out = top_v_out_.has_value()
    ? top_v_out_.value().view({samples, -1})
    : torch::empty({samples, 8},
                   torch::TensorOptions().dtype(torch::kFloat32).device(device));
auto top_k_out = top_k_out_.has_value()
    ? top_k_out_.value().view({samples, -1})
    : torch::empty({samples, 8},
                   torch::TensorOptions().dtype(torch::kInt32).device(device));
```

These tensors are passed to `kimi_k2_sigmoid_top_k_routed_scaled_f32`.

## Observed behavior

We tested fixed logits and gate weights with 1,024 samples, 256 experts, and
top-k = 8. Ten calls produced ten different `top_k` hashes, and all 1,024
samples changed expert selection in at least one call. Sorting the selected
indices did not eliminate the differences.

For 256 experts, valid indices are 0-255. For one sample, repeated executions
returned sets including:

```text
[280, 292, 300, 302, 310, 340, 368, 376]
[268, 294, 302, 324, 342, 350, 362, 368]
[260, 262, 268, 272, 308, 312, 316, 352]
```

For comparison, `torch.topk` returned:

```text
[40, 54, 98, 131, 136, 165, 223, 236]
```

The behavior also occurred with distinct input scores, so it was not ordinary
tie-breaking.

## Proposed fix

Initialize the default output tensors before invoking the cubin:

```diff
- : torch::empty({samples, 8}, ...)
+ : torch::zeros({samples, 8}, ...)
```

With zero-initialized outputs, all returned indices were in range and repeated
calls produced identical outputs. Seven additional test runs covering 371
kernel launches produced no mismatches.

This indicates that the bundled cubin reads or otherwise depends on the initial
contents of its output buffers. Because its implementation is distributed as
an opaque cubin, we have not established the precise internal read-before-write
mechanism. Zero-initializing the wrapper outputs is the immediate verified fix.

## Environment

- NVIDIA B200
- CUDA 12.9
- Tutel commit `96991a970c6287f9a896810941e59578ac257cb3`

A separate test on NVIDIA H100 did not reproduce the issue, so its manifestation
may be architecture-specific.

This issue was identified, analyzed, and documented during a fine-grained,
interactive debugging session with GitHub Copilot, powered by GPT-5.6. The work
was conducted as part of a research project on GPU testing and verification
that is currently under submission.

cc @ghostplant

### PR DESCRIPTION
**Pull request by tyler-utah**
## Summary

- zero-initialize the default Kimi-K2 top-k score output
- zero-initialize the default Kimi-K2 top-k index output
- preserve caller-provided output tensors unchanged

The bundled Kimi-K2 cubin depends on the initial contents of these buffers on
NVIDIA B200. Uninitialized buffers produced nondeterministic expert selections
and, in observed runs, expert indices outside the valid range. Zero-initialized
buffers eliminated both behaviors.

Fixes microsoft/Tutel#339.

## Validation

The fix was validated on NVIDIA B200 across repeated fixed-input executions and
seven additional runs covering 371 kernel launches. The issue does not reproduce
on the available H100 system, so the B200 evidence and reproducer are documented
in the linked issue.

### CODE REVIEW

**Comment by ghostplant** [comment]
Hi. The array is intentionally uninitialized as the immediately following `kimi_k2_sigmoid_top_k_routed_scaled_f32` call is guaranteed to write every element in the valid output range. Zero-initialization therefore provides no correctness benefit, since those elements will all be overwritten anyway. If you encounter some out-of-bounch issues, I believe the root cause lies in somewhere else. Is there any more context?

**Comment by tyler-utah** [comment]
Closing this PR because its premise was incorrect. Our reproducer used 256 experts, while the bundled Kimi-K2 kernel is specialized for 384 experts. With the correct 384-expert input, the existing implementation is deterministic, does not depend on initial output-buffer contents, and matches an independent routing reference.

Zero-initializing the outputs is therefore unnecessary. Sorry for the noise, and thank you for catching this.
